content/blog/zardshard/2023-05-23_how_to_find_memory_leaks.md

```diff
@@ -0,0 +1,59 @@
+
+Both of these concerns can be addressed by using the `leak_analyser.sh` script. This program comes with Haiku by default and can be run directly from the command line with no special setup. First, put the output into a file by running `LD_PRELOAD=libroot_debug.so MALLOC_DEBUG=ges50 program > leaks`. Then, run `leak_analyser.sh leaks`. This will decode the function names and filter out allocations that are normal for a program to exit with. The result will look like
+
+```
+allocation: base: 0x103cf485f60; size: 160; thread: 3130; alignment: 16
+        <libroot.so> operator new(unsigned long) + 0x20
+        <libroot.so> operator new(unsigned long, std::nothrow_t const&) + 0x9
+        <libbe.so> BMenu::AddItem(BMenu*) + 0x27
+        <libbe.so> BMenuField::_AddMenu(BMenu*) + 0x182
+        <libbe.so> BMenuField::_InitMenuBar(BMenu*, BRect, bool) + 0x1ac
+        <libbe.so> BMenuField::BMenuField(BRect, char const*, char const*, BMenu*, unsigned int, unsigned int) + 0xcc
+        <libtracker.so> BPrivate::TFilePanel::Init(BMessage const*) + 0x432
+        <libtracker.so> BPrivate::TFilePanel::TFilePanel(file_panel_mode, BMessenger*, BEntry const*, unsigned int, bool, BMessage*, BRefFilter*, unsigned int, window_look, window_feel, bool) + 0x977
+        <libtracker.so> BFilePanel::BFilePanel(file_panel_mode, BMessenger*, entry_ref const*, unsigned int, bool, BMessage*, BRefFilter*, bool, bool) + 0xe3
+        <_APP_> SavePanel::SavePanel(char const*, BMessenger*, entry_ref*, unsigned int, bool, BMessage*, BRefFilter*, bool, bool) + 0x99
+        <_APP_> IconEditorApp::IconEditorApp() + 0x1f6
+        <_APP_> main + 0x27
+        <_APP_> _start + 0x3f
+        0xf821926ae5 (lookup failed: Invalid Argument)
+        0x7ffefb24d258 (lookup failed: Invalid Argument)
+```
+
+These, then, are two leaks accompanied by a stack trace giving directions to where in the code the memory was allocated. Unfortunately, these do not have line numbers associated with them. For example, `BMenuItem::SetLabel(char const*) + 0x4c` has a `+ 0x4c` instead of a line number. This is the offset to the assembly instruction immediately _after_ the assembly instruction in question.
+
+The Debugger can be used to find what line of source code `BMenuItem::SetLabel(char const*) + 0x4c` is referring to. First, get the disassembly of the function in question. Second, add the offset (`0x4c` in this example) to the address of the first instruction in the function. Third, set a breakpoint to the instruction _before_ the instruction indicated by the offset. Finally, switch back to source code view. The line highlighted by the breakpoint is the line that called the next function in the stack trace or allocated the memory.
```

Comment:
I think you can do this using the `addr2line` command instead. I'm surprised that leak_analyzer doesn't already do that?

Reply:
It doesn't appear the `addr2line` command works. I tried it with a simple "Hello World!" program on Haiku and Linux. Linux gave a file and line number while Haiku gave `??:0`.

Reply:
`llvm-addr2line` works, however, it cannot accept input along the lines of `main+0x3`. I have to first find the base address of the function, e.g. `0x986`, then add `0x3` to it. Let's see if there is any better way.

Reply:
No better way that I could find. I'll just leave the article as written.

Reply:
addr2line only works if the file in question has DWARF information, of course. `libbe` does not by default.

Reply:
Would `libbe` not having DWARF information affect `addr2line` on a "Hello World" program that only uses the C++ standard libraries?

Reply:
The C++ standard libraries also do not have DWARF information, I don't think. Anyway you can likely use `awk` plus `objdump` to turn symbol names into addresses and add them together (or just use a Python script) and then use `llvm-addr2line` on top of that.

Reply:
Indeed, that is the method I had come up with, except I used `nm` and `bc`. Now, I can either (a) create a script to automate this process, (b) put these instructions into the article, or (c) leave the article as written. I had decided to go with option c.

Reply:
I think it makes sense to write a short script and embed it in the article.

Reply:
Good idea!

Reply:
Really this should be fixed in libroot_debug or leak_analyzer, they should be able to show line numbers if Debugger can do it?